File: src/pablo/ghpr.py
"""GitHub PR plumbing via the ``gh`` CLI: CI status, review evaluation,
draft/ready toggling, merge detection.

Review-evaluation rules (spec, "Post-draft state transitions"):
- the PR author's own reviews never count;
- bot reviews are ignored unless whitelisted in ``review.bot_whitelist``;
- only reviews submitted after the PR was last marked ready on GitHub count
  (GitHub's own ``ready_for_review`` timeline event is the anchor — PABLO's
  internal transitions never move it);
- latest review per reviewer wins; changes-requested/comment beats approval;
- GitHub's ``reviewDecision`` is never used.
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from datetime import datetime

from pablo.providers import parse_ts, run_cli

GH_CALL_TIMEOUT_S = 20

CI_FAILURE_CONCLUSIONS = {
    "FAILURE",
    "TIMED_OUT",
    "CANCELLED",
    "ACTION_REQUIRED",
    "STARTUP_FAILURE",
    "ERROR",
}


@dataclass
class PrInfo:
    number: int
    title: str
    state: str  # OPEN | CLOSED | MERGED
    is_draft: bool
    url: str
    merged_at: str | None


def pr_for_branch(repo_slug: str, branch: str) -> PrInfo | None:
    out = run_cli(
        ["gh", "pr", "list", "--repo", repo_slug, "--head", branch,
         "--state", "all", "--json", "number,title,state,isDraft,mergedAt,url",
         "--limit", "1"],
        timeout=GH_CALL_TIMEOUT_S,
    )
    items = json.loads(out)
    if not items:
        return None
    item = items[0]
    return PrInfo(
        number=item["number"],
        title=item["title"],
        state=item["state"],
        is_draft=item["isDraft"],
        url=item["url"],
        merged_at=item.get("mergedAt"),
    )


def prs_for_branches(repo_slug: str, branches: list[str]) -> dict[str, PrInfo]:
    """Same data as calling ``pr_for_branch`` once per branch, but one
    ``gh pr list`` call for the whole repo instead of one per branch."""
    if not branches:
        return {}
    # gh sorts by most-recently-updated, so a generous cap (not just
    # len(branches)) keeps older-but-still-open PRs from falling off the
    # page while still bounding the response for repos with long history.
    limit = min(max(len(branches) * 5, 50), 500)
    out = run_cli(
        ["gh", "pr", "list", "--repo", repo_slug,
         "--state", "all", "--json",
         "number,title,state,isDraft,mergedAt,url,headRefName",
         "--limit", str(limit)],
        timeout=GH_CALL_TIMEOUT_S,
    )
    wanted = set(branches)
    result: dict[str, PrInfo] = {}
    for item in json.loads(out):
        head = item.get("headRefName")
        if head not in wanted or head in result:
            continue
        result[head] = PrInfo(
            number=item["number"],
            title=item["title"],
            state=item["state"],
            is_draft=item["isDraft"],
            url=item["url"],
            merged_at=item.get("mergedAt"),
        )
    return result


def _is_ignored(check: dict, ignore_checks: list[str]) -> bool:
    """Match a check's name (CheckRun) or context (StatusContext) against
    ``ignore_checks`` substrings, case-insensitively.

    Used to exclude manual-gate checks (e.g. a CircleCI approval job nobody
    will click) from CI evaluation — they sit at ``state: PENDING``
    indefinitely (as a StatusContext) or ``conclusion: action_required``
    (as a CheckRun), neither of which ever resolves on its own and would
    otherwise leave the PR stuck pending/red forever.
    """
    if not ignore_checks:
        return False
    name = (check.get("name") or check.get("workflowName") or check.get("context") or "").lower()
    return any(marker.lower() in name for marker in ignore_checks)


def evaluate_ci(rollup: list[dict], ignore_checks: list[str] | None = None) -> str:
    """"green" | "red" | "pending" from a statusCheckRollup list.

    Handles both CheckRun entries (status/conclusion) and StatusContext
    entries (state). No checks at all counts as green — otherwise a project
    without CI could never leave draft. ``ignore_checks`` (from
    ``ci.ignore_checks``) excludes checks whose name/context matches one of
    the given substrings, e.g. a CircleCI approval gate stuck pending
    forever because nobody will click approve.
    """
    ignore_checks = ignore_checks or []
    pending = False
    for check in rollup:
        if _is_ignored(check, ignore_checks):
            continue
        if "state" in check:  # StatusContext
            state = check["state"]
            if state in {"FAILURE", "ERROR"}:
                return "red"
            if state in {"PENDING", "EXPECTED"}:
                pending = True
            continue
        if check.get("status") != "COMPLETED":
            if (check.get("conclusion") or "").upper() in CI_FAILURE_CONCLUSIONS:
                return "red"
            pending = True
            continue
        if (check.get("conclusion") or "").upper() in CI_FAILURE_CONCLUSIONS:
            return "red"
    return "pending" if pending else "green"


def ci_status(repo_slug: str, pr_number: int, ignore_checks: list[str] | None = None) -> str:
    out = run_cli(
        ["gh", "pr", "view", str(pr_number), "--repo", repo_slug,
         "--json", "statusCheckRollup"],
        timeout=GH_CALL_TIMEOUT_S,
    )
    return evaluate_ci(json.loads(out).get("statusCheckRollup") or [], ignore_checks)


_TIMELINE_QUERY = """
query($owner: String!, $repo: String!, $pr: Int!) {
  repository(owner: $owner, name: $repo) {
    pullRequest(number: $pr) {
      createdAt
      timelineItems(itemTypes: [READY_FOR_REVIEW_EVENT], last: 1) {
        nodes { ... on ReadyForReviewEvent { createdAt } }
      }
    }
  }
}
"""


def ready_anchor(repo_slug: str, pr_number: int) -> datetime:
    """When the PR last became ready for review on GitHub.

    Falls back to the PR's createdAt for PRs that were never drafts.
    """
    owner, repo = repo_slug.split("/", 1)
    out = run_cli(
        ["gh", "api", "graphql",
         "-f", f"query={_TIMELINE_QUERY}",
         "-F", f"owner={owner}", "-F", f"repo={repo}", "-F", f"pr={pr_number}"],
        timeout=GH_CALL_TIMEOUT_S,
    )
    pr = json.loads(out)["data"]["repository"]["pullRequest"]
    nodes = [n for n in pr["timelineItems"]["nodes"] if n]
    if nodes:
        return parse_ts(nodes[-1]["createdAt"])
    return parse_ts(pr["createdAt"])


_REVIEWS_QUERY = """
query($owner: String!, $repo: String!, $pr: Int!) {
  repository(owner: $owner, name: $repo) {
    pullRequest(number: $pr) {
      author { login }
      reviews(first: 100) {
        nodes {
          author { login __typename }
          state
          submittedAt
        }
      }
    }
  }
}
"""


def fetch_reviews(repo_slug: str, pr_number: int) -> tuple[str, list[dict]]:
    """(pr_author_login, raw review nodes)."""
    owner, repo = repo_slug.split("/", 1)
    out = run_cli(
        ["gh", "api", "graphql",
         "-f", f"query={_REVIEWS_QUERY}",
         "-F", f"owner={owner}", "-F", f"repo={repo}", "-F", f"pr={pr_number}"],
        timeout=GH_CALL_TIMEOUT_S,
    )
    pr = json.loads(out)["data"]["repository"]["pullRequest"]
    return pr["author"]["login"], [n for n in pr["reviews"]["nodes"] if n]


def evaluate_reviews(
    reviews: list[dict],
    *,
    anchor: datetime,
    author: str,
    bot_whitelist: list[str],
) -> str | None:
    """"approved" | "changes" | None, applying the spec's filtering rules."""
    latest: dict[str, dict] = {}
    for node in reviews:
        review_author = (node.get("author") or {}).get("login")
        if not review_author or review_author == author:
            continue
        is_bot = (
            (node.get("author") or {}).get("__typename") == "Bot"
            or review_author.endswith("[bot]")
        )
        if is_bot and review_author not in bot_whitelist:
            continue
        if node.get("state") not in {"APPROVED", "CHANGES_REQUESTED", "COMMENTED"}:
            continue
        submitted = node.get("submittedAt")
        if not submitted or parse_ts(submitted) <= anchor:
            continue
        current = latest.get(review_author)
        if current is None or parse_ts(submitted) > parse_ts(current["submittedAt"]):
            latest[review_author] = node
    verdicts = {node["state"] for node in latest.values()}
    if {"CHANGES_REQUESTED", "COMMENTED"} & verdicts:
        return "changes"
    if "APPROVED" in verdicts:
        return "approved"
    return None


def mark_ready(repo_slug: str, pr_number: int) -> None:
    run_cli(["gh", "pr", "ready", str(pr_number), "--repo", repo_slug],
            timeout=GH_CALL_TIMEOUT_S)


def mark_draft(repo_slug: str, pr_number: int) -> None:
    run_cli(["gh", "pr", "ready", str(pr_number), "--repo", repo_slug, "--undo"],
            timeout=GH_CALL_TIMEOUT_S)


def is_merged(repo_slug: str, pr_number: int) -> bool:
    out = run_cli(
        ["gh", "pr", "view", str(pr_number), "--repo", repo_slug,
         "--json", "state,mergedAt"],
        timeout=GH_CALL_TIMEOUT_S,
    )
    data = json.loads(out)
    return data["state"] == "MERGED" or bool(data.get("mergedAt"))

File: app/src/Provider/Gh/GhPr.php
<?php

declare(strict_types=1);

namespace Pablo\Provider\Gh;

use Pablo\Domain\Time;
use Pablo\Support\PabloError;
use Pablo\Support\ProcessRunnerInterface;

/**
 * GitHub PR plumbing via the gh CLI: CI status, review evaluation, draft/ready
 * toggling, merge detection.
 */
final class GhPr implements GhPrInterface
{
    public const GH_CALL_TIMEOUT_S = 20;

    private const TIMELINE_QUERY = <<<'GQL'
query($owner: String!, $repo: String!, $pr: Int!) {
  repository(owner: $owner, name: $repo) {
    pullRequest(number: $pr) {
      createdAt
      timelineItems(itemTypes: [READY_FOR_REVIEW_EVENT], last: 1) {
        nodes { ... on ReadyForReviewEvent { createdAt } }
      }
    }
  }
}
GQL;

    private const REVIEWS_QUERY = <<<'GQL'
query($owner: String!, $repo: String!, $pr: Int!) {
  repository(owner: $owner, name: $repo) {
    pullRequest(number: $pr) {
      author { login }
      reviews(first: 100) {
        nodes {
          author { login __typename }
          state
          submittedAt
        }
      }
    }
  }
}
GQL;

    public function __construct(
        private readonly ProcessRunnerInterface $runner,
        private readonly Time $time,
    ) {
    }

    /**
     * @param array<string, mixed> $item
     */
    private function prFromItem(array $item): PrInfo
    {
        return new PrInfo(
            number: (int) $item['number'],
            title: (string) $item['title'],
            state: (string) $item['state'],
            isDraft: (bool) $item['isDraft'],
            url: (string) $item['url'],
            mergedAt: null !== $item['mergedAt'] ? (string) $item['mergedAt'] : null,
            baseRefName: null !== ($item['baseRefName'] ?? null) ? (string) $item['baseRefName'] : null,
        );
    }

    public function prForBranch(string $repoSlug, string $branch): ?PrInfo
    {
        $out = $this->runner->run([
            'gh', 'pr', 'list', '--repo', $repoSlug, '--head', $branch,
            '--state', 'all', '--json', 'number,title,state,isDraft,mergedAt,url,baseRefName',
            '--limit', '1',
        ], timeout: self::GH_CALL_TIMEOUT_S);
        /** @var array<int, array<string, mixed>> $items */
        $items = json_decode($out, true);
        if ([] === $items) {
            return null;
        }

        return $this->prFromItem($items[0]);
    }

    /**
     * @param list<string> $branches
     *
     * @return array<string, PrInfo>
     */
    public function prsForBranches(string $repoSlug, array $branches): array
    {
        if ([] === $branches) {
            return [];
        }
        $limit = min(max(\count($branches) * 5, 50), 500);
        $out = $this->runner->run([
            'gh', 'pr', 'list', '--repo', $repoSlug,
            '--state', 'all', '--json',
            'number,title,state,isDraft,mergedAt,url,headRefName,baseRefName',
            '--limit', (string) $limit,
        ], timeout: self::GH_CALL_TIMEOUT_S);
        $wanted = array_fill_keys($branches, true);
        $result = [];
        /** @var array<int, array<string, mixed>> $items */
        $items = json_decode($out, true);
        foreach ($items as $item) {
            $head = $item['headRefName'] ?? null;
            if (null === $head || !isset($wanted[$head]) || isset($result[$head])) {
                continue;
            }
            $result[$head] = $this->prFromItem($item);
        }

        return $result;
    }

    /**
     * @param list<array{0: string, 1: list<string>}> $repoBranches [slug, branches]
     *
     * @return array<string, array<string, PrInfo>> slug => branch => PrInfo
     */
    public function prsForBranchesBulk(array $repoBranches): array
    {
        if ([] === $repoBranches) {
            return [];
        }

        $commands = [];
        $slugIndex = [];
        $branchesByIndex = [];
        foreach ($repoBranches as $i => [$slug, $branches]) {
            $slugIndex[$i] = $slug;
            $branchesByIndex[$i] = $branches;
            $limit = min(max(\count($branches) * 5, 50), 500);
            $commands[] = [
                'gh', 'pr', 'list', '--repo', $slug,
                '--state', 'all', '--json',
                'number,title,state,isDraft,mergedAt,url,headRefName,baseRefName',
                '--limit', (string) $limit,
            ];
        }

        $results = $this->runner->runParallel($commands, check: false, timeout: self::GH_CALL_TIMEOUT_S);
        $all = [];
        foreach ($results as $i => $out) {
            $slug = $slugIndex[$i];
            $branches = $branchesByIndex[$i];
            $wanted = array_fill_keys($branches, true);
            $all[$slug] = [];
            try {
                /** @var array<int, array<string, mixed>> $items */
                $items = json_decode($out, true, 512, \JSON_THROW_ON_ERROR);
            } catch (\Throwable) {
                continue;
            }
            foreach ($items as $item) {
                $head = $item['headRefName'] ?? null;
                if (null === $head || !isset($wanted[$head]) || isset($all[$slug][$head])) {
                    continue;
                }
                $all[$slug][$head] = $this->prFromItem($item);
            }
        }

        return $all;
    }

    /**
     * @param list<string> $ignoreChecks
     */
    private function isIgnored(CiCheck $check, array $ignoreChecks): bool
    {
        foreach ($ignoreChecks as $marker) {
            if (str_contains($check->name, strtolower((string) $marker))) {
                return true;
            }
        }

        return false;
    }

    /**
     * @param list<CiCheck> $rollup
     * @param list<string>  $ignoreChecks
     */
    public function evaluateCi(array $rollup, array $ignoreChecks = []): string
    {
        $pending = false;
        foreach ($rollup as $check) {
            if ($this->isIgnored($check, $ignoreChecks)) {
                continue;
            }
            if ($check->failed) {
                return 'red';
            }
            if ($check->pending) {
                $pending = true;
            }
        }

        return $pending ? 'pending' : 'green';
    }

    public function ciStatus(string $repoSlug, int $prNumber, array $ignoreChecks = []): string
    {
        $out = $this->runner->run([
            'gh', 'pr', 'view', (string) $prNumber, '--repo', $repoSlug,
            '--json', 'statusCheckRollup',
        ], timeout: self::GH_CALL_TIMEOUT_S);
        /** @var array<string, mixed> $data */
        $data = json_decode($out, true);
        $rollup = array_values(array_map(CiCheck::fromRollup(...), $data['statusCheckRollup'] ?? []));

        return $this->evaluateCi($rollup, $ignoreChecks);
    }

    public function readyAnchor(string $repoSlug, int $prNumber): \DateTimeImmutable
    {
        $ownerRepo = explode('/', $repoSlug, 2);
        $out = $this->runner->run([
            'gh', 'api', 'graphql',
            '-f', 'query='.self::TIMELINE_QUERY,
            '-F', 'owner='.$ownerRepo[0], '-F', 'repo='.$ownerRepo[1], '-F', 'pr='.$prNumber,
        ], timeout: self::GH_CALL_TIMEOUT_S);
        /** @var array<string, mixed> $data */
        $data = json_decode($out, true);
        $pr = $data['data']['repository']['pullRequest'];
        $nodes = array_values(array_filter($pr['timelineItems']['nodes'] ?? [], static fn ($n) => null !== $n));
        if ([] !== $nodes) {
            return $this->time->parseTs((string) $nodes[\count($nodes) - 1]['createdAt']);
        }

        return $this->time->parseTs((string) $pr['createdAt']);
    }

    public function fetchReviews(string $repoSlug, int $prNumber): PrReviews
    {
        $ownerRepo = explode('/', $repoSlug, 2);
        $out = $this->runner->run([
            'gh', 'api', 'graphql',
            '-f', 'query='.self::REVIEWS_QUERY,
            '-F', 'owner='.$ownerRepo[0], '-F', 'repo='.$ownerRepo[1], '-F', 'pr='.$prNumber,
        ], timeout: self::GH_CALL_TIMEOUT_S);
        /** @var array<string, mixed> $data */
        $data = json_decode($out, true);
        $pr = $data['data']['repository']['pullRequest'];
        $nodes = array_values(array_filter($pr['reviews']['nodes'] ?? [], static fn ($n) => null !== $n));

        return new PrReviews(
            (string) $pr['author']['login'],
            array_map(ReviewNode::fromNode(...), $nodes),
        );
    }

    /**
     * @param list<ReviewNode> $reviews
     * @param list<string>     $botWhitelist
     */
    public function evaluateReviews(array $reviews, \DateTimeImmutable $anchor, string $author, array $botWhitelist): ?string
    {
        $latest = [];
        foreach ($reviews as $node) {
            $reviewAuthor = $node->login;
            if ('' === $reviewAuthor || $reviewAuthor === $author) {
                continue;
            }
            if ($node->isBot() && !\in_array($reviewAuthor, $botWhitelist, true)) {
                continue;
            }
            if (!\in_array($node->state, ['APPROVED', 'CHANGES_REQUESTED', 'COMMENTED'], true)) {
                continue;
            }
            $submitted = $node->submittedAt;
            if (null === $submitted || $this->time->parseTs($submitted) <= $anchor) {
                continue;
            }
            $current = $latest[$reviewAuthor] ?? null;
            if (null === $current || $this->time->parseTs($submitted) > $this->time->parseTs((string) $current->submittedAt)) {
                $latest[$reviewAuthor] = $node;
            }
        }
        $verdicts = [];
        foreach ($latest as $node) {
            $verdicts[$node->state] = true;
        }
        if (isset($verdicts['CHANGES_REQUESTED']) || isset($verdicts['COMMENTED'])) {
            return 'changes';
        }
        if (isset($verdicts['APPROVED'])) {
            return 'approved';
        }

        return null;
    }

    public function markReady(string $repoSlug, int $prNumber): void
    {
        $this->runner->run(['gh', 'pr', 'ready', (string) $prNumber, '--repo', $repoSlug], timeout: self::GH_CALL_TIMEOUT_S);
    }

    public function markDraft(string $repoSlug, int $prNumber): void
    {
        $this->runner->run(['gh', 'pr', 'ready', (string) $prNumber, '--repo', $repoSlug, '--undo'], timeout: self::GH_CALL_TIMEOUT_S);
    }

    public function isMerged(string $repoSlug, int $prNumber): bool
    {
        $out = $this->runner->run([
            'gh', 'pr', 'view', (string) $prNumber, '--repo', $repoSlug,
            '--json', 'state,mergedAt',
        ], timeout: self::GH_CALL_TIMEOUT_S);
        /** @var array<string, mixed> $data */
        $data = json_decode($out, true);

        return 'MERGED' === $data['state'] || !empty($data['mergedAt']);
    }

    /** @return array<int, string> */
    public function rerunCi(string $repoSlug, string $branch): array
    {
        $out = $this->runner->run([
            'gh', 'run', 'list', '--repo', $repoSlug,
            '--branch', $branch, '--status', 'completed',
            '--limit', '20', '--json', 'databaseId,workflowName',
        ], timeout: self::GH_CALL_TIMEOUT_S);
        /** @var array<int, array<string, mixed>> $runs */
        $runs = json_decode($out, true);
        if ([] === $runs) {
            throw new PabloError("no completed workflow runs found on branch {$branch}");
        }
        $seen = [];
        $latest = [];
        foreach ($runs as $run) {
            $wf = $run['workflowName'];
            if (!isset($seen[$wf])) {
                $seen[$wf] = true;
                $latest[] = $run;
            }
        }
        $rerunIds = [];
        foreach ($latest as $run) {
            $runId = (string) $run['databaseId'];
            $this->runner->run(['gh', 'run', 'rerun', $runId, '--repo', $repoSlug], timeout: self::GH_CALL_TIMEOUT_S);
            $rerunIds[] = $runId;
        }

        return $rerunIds;
    }
}
